source/reference/array.rst

```diff
@@ -236,7 +236,200 @@ Unevaluated Items
 
 |draft2019-09|
 
-Documentation Coming Soon
+The ``unevaluatedItems`` keyword selects any data types not evaluated
+by an ``items``, ``prefixItems``, or `contains` keyword. Just as
+unevaluated"properties" affect only "properties" in an object, only
+"item"-related keywords affect unevaluated"items".
+
+It also applies inside valid subschemas with these keywords:
+- ``allOf``
+- ``anyOf``
+- ``oneOf``
+- ``not``
+- ``if``
+- ``then``
+- ``else``
+
+The main reason to use this keyword is to extend an array with extra
+arguments.
+
+For this first example, we'll use ``unevaluatedItems`` to select any
```

Comment:
I've edited all the other things on this page, but could you clarify what you mean here? My understanding is that ``unevaluatedItems`` works like ``unevaluatedProperties`` in that it _does_ recognize items declared in subschemas, whereas ``additionalProperties`` fails validation when attempting to extend a schema using ``allOf``. (I got this from the "Object" page.) For instance, I made this, which I validated with the schema validator:

```
{
  "allOf": [
    {
      "type": "array",
      "items": {
        "quantity": "number",
        "product": "string",
      },
    }
  ],

  "items": {
    "quantity": { "enum": ["1", "2", "3"] }
  },
    "required": ["quantity"],
  "unevaluatedItems": false
}
```
Or did I misunderstand something? You're the expert.

Reply:
> unevaluatedItems works like unevaluatedProperties in that it does recognize items declared in subschemas, whereas additionalProperties fails validation when attempting to extend a schema using allOf

That's exactly right -- and your example I *think* maybe is fine, but what instances (data) were you going to use for that?

--

Regardless, what I meant is something like the following (sorry for being terse both previously and possibly still :)).

Imagine you have lists of numbers like `[1, 2, 2, 2, 2]`, `[1, 2, 2]`, `[1, 2]` -- so, a list where the first number is 1 and then you are allowed any number of elements after but they all have to be the number `2`.

So you write this schema:

```json
{
    "prefixItems": [{"const": 1}],
    "items": {"const": 2},
}
```

great, it works, lists that look like the above are valid, other lists are not.

Now you actually want to allow lists either like the ones above, or ones where the first number is `"a"`. OK great, let's modify the schema to also allow `"a"`

```json
{
    "oneOf": [{"prefixItems": [{"const": 1}]}, {"prefixItems": [{"const": "a"}]}],
    "items": {"const": 2},
}
```

because that's what `oneOf` does, right? Except no, that doesn't work! The reason is that `items` will now apply to *all* of the elements, requiring everything to be `2`. And indeed as you say `prefixItems` is now somehow not "hiding" elements from `items` the way it did in the simplest example.

However, `unevaluatedItems` indeed does what we want! If we just change `items` to `unevaluatedItems`, our schema does what we want -- lists starting with `[1, 2, ...]` are valid as before, as are `["a", 2, 2, ...]`, and others are invalid.

Now, my example is slightly contrived because you could also write the schema as:

```json
{
    "prefixItems": [{"oneOf": [{"const": 1}]}, {"const": "a"}]}],
    "items": {"const": 2},
}
```

and this would indeed work even with `items`, but I think this kind of example is clearer to start with.

Does that help at least explain what I mean (even if we don't necessarily change anything and stick with what you have)?

Reply:
Ah, okay. So the ``oneOf`` (I assume ``anyOf`` and ``allOf`` work the same way) is saying "one of either 1, a, or 2" BUT ALSO all items must be 2, therefore it fails if anything is not a 2. The ``xxxOf`` attaches itself to the ``items``, but ``unevaluatedItems`` are not evaluated by the ``xxxOf``.

I get it now, though this is an entirely different direction than I went. I did need an example of when ``uneavluatedItems`` is something other than a Boolean though, so I'll incorporate this. It's going to be a lot of examples under one subheader.